File: src/parsing/operations/order_by.ts
import Expression from "../expressions/expression";
import Reference from "../expressions/reference";
import Operation from "./operation";

export interface SortField {
    direction: "asc" | "desc";
    /** The parsed expression to evaluate for this sort field. */
    expression: Expression;
}

/**
 * Represents an ORDER BY operation that sorts results.
 *
 * Can be attached to a RETURN operation (sorting its results),
 * or used as a standalone accumulating operation after a non-aggregate WITH.
 *
 * Supports both simple field references and arbitrary expressions:
 * @example
 * ```
 * RETURN x ORDER BY x DESC
 * RETURN x ORDER BY toLower(x.name) ASC
 * RETURN x ORDER BY string_distance(toLower(x.name), toLower('Thomas')) ASC
 * ```
 */
class OrderBy extends Operation {
    private _fields: SortField[];
    private _results: Record<string, any>[] = [];
    /**
     * Parallel array of pre-computed sort-key tuples, one entry per
     * accumulated result row.  Each inner array has one value per sort
     * field, in the same order as `_fields`.
     */
    private _sortKeys: any[][] = [];

    constructor(fields: SortField[]) {
        super();
        this._fields = fields;
    }

    public get fields(): SortField[] {
        return this._fields;
    }

    /**
     * Evaluates every sort-field expression against the current runtime
     * context and stores the resulting values.  Must be called once per
     * accumulated row (from `Return.run()`).
     */
    public captureSortKeys(): void {
        this._sortKeys.push(this._fields.map((f) => f.expression.value()));
    }

    /**
     * Sorts an array of records using the pre-computed sort keys captured
     * during accumulation.  When no keys have been captured (e.g.
     * aggregated returns), falls back to looking up simple reference
     * identifiers in each record.
     */
    public sort(records: Record<string, any>[]): Record<string, any>[] {
        const useKeys = this._sortKeys.length === records.length;
        // Build an index array so we can sort records and keys together.
        const indices = records.map((_, i) => i);
        const keys = this._sortKeys;

        // Pre-compute fallback field names for when sort keys aren't
        // available (aggregated returns).  Simple references like `x`
        // map to the column name; complex expressions have no fallback.
        const fallbackFields: (string | null)[] = this._fields.map((f) => {
            const root = f.expression.firstChild();
            if (root instanceof Reference && f.expression.childCount() === 1) {
                return (root as Reference).identifier;
            }
            return null;
        });

        indices.sort((ai, bi) => {
            for (let f = 0; f < this._fields.length; f++) {
                let aVal: any;
                let bVal: any;
                if (useKeys) {
                    aVal = keys[ai][f];
                    bVal = keys[bi][f];
                } else if (fallbackFields[f] !== null) {
                    aVal = records[ai][fallbackFields[f]!];
                    bVal = records[bi][fallbackFields[f]!];
                } else {
                    continue;
                }
                let cmp = 0;
                if (aVal == null && bVal == null) cmp = 0;
                else if (aVal == null) cmp = -1;
                else if (bVal == null) cmp = 1;
                else if (aVal < bVal) cmp = -1;
                else if (aVal > bVal) cmp = 1;
                if (cmp !== 0) {
                    return this._fields[f].direction === "desc" ? -cmp : cmp;
                }
            }
            return 0;
        });
        return indices.map((i) => records[i]);
    }

    /**
     * When used as a standalone operation (after non-aggregate WITH),
     * accumulates records to sort later.
     */
    public async run(): Promise<void> {
        const record: Record<string, any> = {};
        // Collect current variable values from the context
        // This gets called per-row, and then finish() sorts and emits
        await this.next?.run();
    }

    public async initialize(): Promise<void> {
        this._results = [];
        this._sortKeys = [];
        await this.next?.initialize();
    }

    public get results(): Record<string, any>[] {
        return this._results;
    }
}

export default OrderBy;

File: flowquery-py/src/parsing/operations/order_by.py
"""Represents an ORDER BY operation that sorts results."""

import functools
from typing import TYPE_CHECKING, Any, Dict, List, Optional

from .operation import Operation

if TYPE_CHECKING:
    from ..expressions.expression import Expression


class SortField:
    """A single sort specification: expression and direction."""

    def __init__(self, expression: 'Expression', direction: str = "asc"):
        self.expression = expression
        self.direction = direction


class OrderBy(Operation):
    """Represents an ORDER BY operation that sorts results.

    Can be attached to a RETURN operation (sorting its results),
    or used as a standalone accumulating operation after a non-aggregate WITH.

    Supports both simple field references and arbitrary expressions:

    Example::

        RETURN x ORDER BY x DESC
        RETURN x ORDER BY toLower(x.name) ASC
        RETURN x ORDER BY string_distance(toLower(x.name), toLower('Thomas')) ASC
    """

    def __init__(self, fields: List[SortField]):
        super().__init__()
        self._fields = fields
        self._results: List[Dict[str, Any]] = []
        self._sort_keys: List[List[Any]] = []

    @property
    def fields(self) -> List[SortField]:
        return self._fields

    def capture_sort_keys(self) -> None:
        """Evaluate every sort-field expression against the current runtime
        context and store the resulting values.  Must be called once per
        accumulated row (from ``Return.run()``)."""
        self._sort_keys.append([f.expression.value() for f in self._fields])

    def sort(self, records: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Sort records using pre-computed sort keys captured during
        accumulation.  When no keys have been captured (e.g. aggregated
        returns), falls back to looking up simple reference identifiers
        in each record."""
        from ..expressions.reference import Reference

        use_keys = len(self._sort_keys) == len(records)
        keys = self._sort_keys

        # Pre-compute fallback field names for when sort keys aren't
        # available (aggregated returns).
        fallback_fields: List[Optional[str]] = []
        for f in self._fields:
            root = f.expression.first_child()
            if isinstance(root, Reference) and f.expression.child_count() == 1:
                fallback_fields.append(root.identifier)
            else:
                fallback_fields.append(None)

        indices = list(range(len(records)))

        def compare(ai: int, bi: int) -> int:
            for f_idx, sf in enumerate(self._fields):
                if use_keys:
                    a_val = keys[ai][f_idx]
                    b_val = keys[bi][f_idx]
                elif fallback_fields[f_idx] is not None:
                    a_val = records[ai].get(fallback_fields[f_idx])  # type: ignore[arg-type]
                    b_val = records[bi].get(fallback_fields[f_idx])  # type: ignore[arg-type]
                else:
                    continue
                cmp = 0
                if a_val is None and b_val is None:
                    cmp = 0
                elif a_val is None:
                    cmp = -1
                elif b_val is None:
                    cmp = 1
                elif a_val < b_val:
                    cmp = -1
                elif a_val > b_val:
                    cmp = 1
                if cmp != 0:
                    return -cmp if sf.direction == "desc" else cmp
            return 0

        indices.sort(key=functools.cmp_to_key(compare))
        return [records[i] for i in indices]

    async def run(self) -> None:
        """When used as a standalone operation, passes through to next."""
        if self.next:
            await self.next.run()

    async def initialize(self) -> None:
        self._results = []
        self._sort_keys = []
        if self.next:
            await self.next.initialize()

    @property
    def results(self) -> List[Dict[str, Any]]:
        return self._results
